### PR TITLE
[serve] remove support for nested `DeploymentResponses`

### LINKED ISSUE #46428: [core][serve] 1MB latency performance regression
### What happened + What you expected to happen

Around 6/14, latency for sending a request with 1MB payload through a serve `DeploymentHandle` increased from ~3.4s to ~4.6s. 

Screenshot 2024-07-03 at 3 58 58 PM

From bisecting, https://github.com/ray-project/ray/commit/d729815c4b88232dcb20860ff5ee1e7f871111f4 seems to be the offending commit.

### Versions / Dependencies

n/a

### Reproduction script

Run `python release/serve_tests/workloads/microbenchmarks.py`.

### Issue Severity

None

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Remove support for passing `DeploymentResponses` in nested objects to downstream serve deployment handle calls, e.g:
```
handle.remote({"arg": other_handle.remote()})
```

This doesn't affect the no-op latency, but improves latencies of requests that carry a large payload.

Current handle latencies
Screenshot 2024-08-19 at 5 28 17 PM

New handle latencies:
```
{
  "handle_p50_latency": 1.7133924999939154,
  "handle_1mb_p50_latency": 2.931859499994971,
  "handle_10mb_p50_latency": 12.14768799999888,
}
```

## Related issue number

closes https://github.com/ray-project/ray/issues/46428

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@zcin let's make sure a good error message is raised if someone tries to pass a nested deployment response. You can do this by defining a `__reduce__` method and raising an informative exception there.

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -437,60 +436,60 @@ def update_deployment_config(self, deployment_config: DeploymentConfig):
...
+
+        kwarg_tasks = []
+        response_keys = []
+        for k, obj in request_kwargs.items():
+            if isinstance(obj, DeploymentResponseGenerator):
+                raise RuntimeError(
+                    "Streaming deployment handle results cannot be passed to "
+                    "downstream handle calls. If you have a use case requiring "
+                    "this feature, please file a feature request on GitHub."
+                )
```
Comment (edoakes) [inline python/ray/serve/_private/router.py:473]: nit: share this error message to avoid copypasta

Comment (zcin) [reply python/ray/serve/_private/router.py:473]: done

**Comment by zcin** [comment]
> @zcin let's make sure a good error message is raised if someone tries to pass a nested deployment response. You can do this by defining a `__reduce__` method and raising an informative exception there.

@edoakes Added an informative error message in `__reduce__` of `DeploymentResponse`, PTAL!

**Comment by edoakes** [comment]
> > @zcin let's make sure a good error message is raised if someone tries to pass a nested deployment response. You can do this by defining a `__reduce__` method and raising an informative exception there.
> 
> @edoakes Added an informative error message in `__reduce__` of `DeploymentResponse`, PTAL!

Let's add a test for it too, then LGTM

**Review by edoakes — APPROVED** [review]

**Pull request merged**